### PR TITLE
Add a simple set of Linq benchmarks

### PR DESCRIPTION
**Pull request by AndyAyersMS**
@dotnet/jit-contrib PTAL.

### CODE REVIEW

**Comment by BruceForstall** [comment]
LGTM

**Comment by AndyAyersMS** [comment]
@dotnet-bot retest Ubuntu x64 Release Build and Test

**Comment by GSPP** [comment]
I don't think orderby, groupby and join are things where the JIT can do much. The underlying algorithms are too big to ever be inlined or somehow transformed.

I think the scope of the benchmarks should be Select, Where, Sum, ToList, ToArray, foreach. These are, I think, the ones with realistic optimization opportunities (and significant ones, like an order of magnitude).

**Comment by AndyAyersMS** [comment]
This is just a first draft that we'll undoubtedly update as we dig in deeper. For example I think this initial cut probably stresses comparisons of decimals a bit too much; not much the jit can do about these.

I added an `orderby` since I want to see if the sort it uses gets reasonably well optimized (at least as well as the traditional array sort).

Feel free to augment this with more or better examples.

**Comment by jods4** [comment]
@AndyAyersMS 
It's good to start somewhere... 

For an initial feedback I'd say: 

1. There should be a few longer chains of LINQ operators. Like `.Select().Where().Where().Select().Sum()`.
It's not totally unrealistic: first `Select` could be a `let`, dynamically adding conditions results in repeated `Where`, the last `Select` is to fetch the value you wan to sum only.
Long chains show how much loop merging could help (all this could be a single loop).

2. I agree that complex operators that need to buffer data (e.g. `orderby`, `join`, `groupby`) offer less optimization opportunities, but a few cases would be good to check how the mix with the other optimizations.

3. You probably need to test against various sources: `IEnumerable<T>`, `IList<T>`, `T[]`. Could be done by running the same tests on the same data but with a `ToList()`, `ToArray()` call.

4. In the long run we should include as many operations as possible, so that: optimization opportunities are not missed, those methods don't break the other optimizations (e.g. break loop merging). A few examples off the top of my head: 
`let` or any intermediate `Select()`. Especially since those usually allocate memory that could be local variables.
`Take()`, `Skip()`. With extra optimization opportunities when the target is known to be an array or `ICollection` (i.e. has an indexer).
More "collectors", such as `ToList()`, `First()`, `At()`, `Any()`, `All()`, `Sum()`, `Aggregate()`, which all have their own optimization opportunities and patterns.

Hum.... that's a lot. I guess it's good to start small <_<

File: tests/src/JIT/Performance/CodeQuality/Linq/Linq.cs [hunk 1]
```diff
@@ -0,0 +1,594 @@
...
+
+    public static List<Product> GetProductList()
+    {
+        if (s_productList == null)
+            createLists();
+
+        return s_productList;
+    }
+
+    private static void createLists()
```
Comment (pgavlin) [inline tests/src/JIT/Performance/CodeQuality/Linq/Linq.cs]: Nit: should be `CreateLists` as per style guidelines.

File: tests/src/JIT/Performance/CodeQuality/Linq/Linq.cs [hunk 2]
```diff
@@ -0,0 +1,594 @@
...
+    public const int IterationsOrder00 = 25000;
+#endif
+
+    private static volatile object s_volatileObject;
+
+    private static void Escape(object obj)
+    {
+        s_volatileObject = obj;
+    }
+    [MethodImpl(MethodImplOptions.NoInlining)]
```
Comment (pgavlin) [inline tests/src/JIT/Performance/CodeQuality/Linq/Linq.cs]: Nit: formatting is off--looks like the blank line below should instead be above.

**Comment by pgavlin** [comment]
LGTM aside from a couple naming/formatting nits.

**Comment by CarolEidt** [comment]
LGTM. Over time, one might want to add both additional data sets and entries, so it might be good to consider (in future) validating the results by computing them without LINQ (instead of hard-coding), just to make it more adaptable and extensible.

File: tests/src/JIT/Performance/CodeQuality/Linq/Linq.cs (hunk 2 above)
Comment (AndyAyersMS) [inline tests/src/JIT/Performance/CodeQuality/Linq/Linq.cs]: Odd cause I ran the codeformatter over this. Will fix.

**Comment by AndyAyersMS** [comment]
@pgavlin updated...

**Comment by AndyAyersMS** [comment]
Ubuntu failure looks unrelated (perhaps a flaky test). Will retry...

@dotnet-bot retest Ubuntu x64 Release Build and Test

**Pull request merged**